refactor movie repository implementation

MustafaIbrahim96 opened the pull request.

MustafaIbrahim96 committed:
add movie similar data source and refactor movie repository

MustafaIbrahim96 committed:
Merge remote-tracking branch 'origin/feature/details-ui' into refactor/local-medioa-details

MustafaIbrahim96 committed:
add movie similar data source and refactor movie repository

MustafaIbrahim96 committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into refactor/local-medioa-details

MustafaIbrahim96 committed:
fix cast and reviews movies in repository

MustafaIbrahim96 committed:
Merge branch 'develop' into refactor/local-medioa-details

MustafaIbrahim96 committed:
refactor movieRepository test cases

MustafaIbrahim96 committed:
remove app release dm

mohammed-akkad commented on repository/movie/src/main/java/com/repository/movie/repository/MovieRepositoryImpl.kt:
```diff
@@ -5,77 +5,166 @@ import com.domain.mediaDetails.exception.NoInternetConnectionException
+            } else {
+                val remoteMovie = movieDetailsRemoteDataSource.getMovieDetails(movieId, language)
+                movieLocalDataSource.addMovie(remoteMovie.toLocalDto(language))
+                remoteMovie.toEntity()
```
should get data from local after add it form remote

Joseph-Sameh-0 replied:
there is tow approaches when the data is not in the local
- add the data in the local and map the returned data from the remote to domain
- add the data in the local and always get the data from the local and map it to the domain
the two approaches is valid

mohammed-akkad commented on datasource/local/movie/src/main/java/com/datasource/local/dao/MovieSimilarDao.kt:
```diff
@@ -0,0 +1,17 @@
+    suspend fun addSimilarMovies(movies: List<MovieSimilarEntity>)
+
+    @Query("SELECT * FROM movies_similar_table WHERE id = :movieId AND page = :page AND language = :language")
+    suspend fun getSimilarMovies(movieId: Int,page: Int, language: String): List<MovieSimilarEntity>?
```
you can make returen type list of movie simillar without nullable becuse if it null return empty list

mohammed-akkad commented on repository/movie/src/main/java/com/repository/movie/mapper/MovieMapper.kt:
```diff
@@ -114,9 +117,13 @@ fun MovieLogoDto.toEntity(id: Int): Image {
 fun MovieReviewDto.toEntity(): Review {
     return Review(
+        createdAt = try {
+            LocalDate.parse(this.createdAt.orEmpty())
```
you can edit this to be like this 
```suggestion
            LocalDate.parse(this.createdAt ?: "9999,1,1")
```

Joseph-Sameh-0 replied:
Sometimes the API returns emty string this will not handle it

Joseph-Sameh-0 committed:
Sometimes the API returns emty string
this will not handle it

Co-authored-by: Mohammed Al-Akkad <93793786+mohammed-akkad@users.noreply.github.com>

Joseph-Sameh-0 commented on repository/movie/src/main/java/com/repository/movie/mapper/MovieMapper.kt:
```diff
@@ -114,9 +117,13 @@ fun MovieLogoDto.toEntity(id: Int): Image {
         id = this.id.orEmpty(),
         name = this.authorDetails?.name.orEmpty(),
-        createdAt = LocalDate.parse(this.createdAt?.substring(0,10) ?: "2025-01-01"),
         avatarUrl = this.authorDetails?.avatarPath.toImageUrl().orEmpty(),
```
the createdAt in the API is Date time
make substring(0, 10)

or make the type DateTime (recommended)

Joseph-Sameh-0 committed:
Revert "Sometimes the API returns emty string"

This reverts commit e8584ef2ee67f05c0a9ca663af74c0ff4f837490.

Joseph-Sameh-0 committed:
refactor: Ensure Non-Null Return Types for Local Data Sources

This commit modifies local data source interfaces and their implementations (`MovieCastLocalDataSource`, `MovieSimilarLocalDataSource`) along with their respective DAOs (`MovieCastDao`, `MovieSimilarDao`, `MovieReviewDao`) to return non-nullable lists (`List<Type>`) instead of nullable lists (`List<Type>?`).

This change simplifies null handling in the `MovieRepositoryImpl` by removing unnecessary null checks (`isNullOrEmpty()`) and directly using `isNotEmpty()`.

Additionally, the `MovieMapper.kt` has been updated to correctly parse the `createdAt` date string in `MovieReviewDto.toEntity()` by taking a substring to ensure only the date part is parsed.

The following specific changes were made:
- Updated `MovieCastDao`, `MovieSimilarDao`, and `MovieReviewDao` query methods to return `List<Entity>` instead of `List<Entity>?`.
- Corrected a typo in the `MovieSimilarDao` query from `id` to `movieId`.
- Updated `MovieCastLocalDataSource` and `MovieSimilarLocalDataSource` interface methods and their implementations (`MovieCastLocalDataSourceImp`, `MovieSimilarLocalDataSourceImp`) to reflect the non-null return types.
- Simplified null checks in `MovieRepositoryImpl` for `localCast` and `localMoviesSimilar` by using `isNotEmpty()` directly.
- Modified `MovieReviewDto.toEntity()` in `MovieMapper.kt` to parse only the date portion of `createdAt` by using `substring(0,10)`.

Joseph-Sameh-0 committed:
refactor: Rename movieDao to tvShowDao in TvShowDetailDataBase

This commit renames the `movieDao()` method to `tvShowDao()` in the `TvShowDetailDataBase` class for clarity and consistency. The corresponding dependency injection in `RoomModule.kt` has also been updated.

Joseph-Sameh-0 committed:
Update MovieRepositoryImplTest.kt

Joseph-Sameh-0 committed:
update release-info.txt
